src/puppetlabs/puppetdb/middleware.clj

```diff
@@ -286,10 +286,30 @@
           (app (assoc req :body-string payload))
           (http/error-response (str "Missing required parameter 'payload'")))
         "application/json"
-        (let [body-string (request/body-string req)]
-          (if (nil? body-string)
+        (let [{:strs [certname version command]} params
+              body-string (request/body-string req)
+              param-post? (boolean (and certname version command))]
+
+          (cond
+            (and (not param-post?) (or certname version command))
+            (http/error-response (str "Command submission via POST parameters requires certname, version, and command."))
```

Comment:
Also wondered about a possibly more direct message like "Commands submitted via POST require certname, version, and command parameters." or similar.